src/Database/LSMTree/Internal/IndexOrdinary.hs

```diff
@@ -0,0 +1,222 @@
+-- | A general-purpose fence pointer index.
+module Database.LSMTree.Internal.IndexOrdinary
+(
+    IndexOrdinary (IndexOrdinary),
+
+    -- * Search
+    search,
+
+    -- * Deserialisation
+    fromSBS
+)
+where
+
+import Prelude hiding (length, drop, takeWhile)
+
+import Control.Monad
+       (
+           when
+       )
+import Data.Word
+       (
+           Word8,
+           Word16,
+           Word32,
+           Word64,
+           byteSwap32
+       )
+import Data.Primitive.ByteArray
+       (
+           ByteArray,
+           indexByteArray
+       )
+import Data.ByteString.Short
+       (
+           ShortByteString (ShortByteString)
+       )
+import Data.ByteString.Short qualified as ShortByteString
+       (
+           length
+       )
+import Data.Vector
+       (
+           Vector,
+           fromList,
+           length,
+           (!),
+           drop,
+           takeWhile,
+           findIndexR
+       )
+import Data.Vector.Primitive qualified as Primitive
+       (
+           Vector (Vector),
+           force,
+           null,
+           length,
+           take,
+           drop,
+           splitAt,
+       )
+import Database.LSMTree.Internal.Serialise
+       (
+           SerialisedKey (SerialisedKey')
+       )
+import Database.LSMTree.Internal.Entry
+       (
+           NumEntries (NumEntries)
+       )
+import Database.LSMTree.Internal.IndexCompact
+       (
+           PageNo (PageNo),
+           PageSpan (PageSpan)
+       )
+
+{-|
+    The version of the index and its serialisation format that is supported by
+    this module.
+-}
+supportedVersion :: Word32
+supportedVersion = 1
+
+{-|
+    A general-purpose fence pointer index.
+
+    An index is represented by a vector that maps the number of each page to the
+    key stored last in this page or, if the page is an overflow page, to the key
+    of the corresponding key–value pair. From this, the following constraints on
+    the vector follow:
+
+      * The vector must be non-empty.
+
+      * The elements of the vector must be non-decreasing.
+-}
+newtype IndexOrdinary = IndexOrdinary (Vector SerialisedKey)
+
+-- * Search
+
+{-|
+    Returns the span of pages that contain the key–value pair of a certain key
+    if there is such a pair, or an arbitrary, valid page span if there is not.
+-}
+search :: SerialisedKey -> IndexOrdinary -> PageSpan
+search key (IndexOrdinary lastKeys) = PageSpan (PageNo start) (PageNo end) where
+
+    start :: Int
+    start | protoStart < pageCount = protoStart
+          | otherwise              = maybe 0 succ $
```

Comment:
Hmm, what I see in some other modules looks like a mix of 2-space indentation and 4-space indentation to me. To some extend, this is because of an indentation style that partly uses two indentations for marking one logical level, but in other places I even cannot see why 4 spaces are used.

My opinion is that 4-space indentation makes the indentation structure more easily recognizable and is not really a problem if you avoid those two indentations for one logical level. Therefore, I’d like to keep it, but if there is a strong urge to have this module use 2-space indentation I may change the indentation depth.

Reply:
You're right, we use a mix: I was under the impression most of us had used 2-space indentation as the baseline. I'm not really sure what you mean by "... is not really a problem if you avoid those two indentations for one logical level". Could you elaborate?

IMO 2-space indentation saves a lot of horizontal real estate in source code, and that's why I prefer it. For example, [these lines](https://github.com/IntersectMBO/lsm-tree/pull/244/files#diff-d9780d09d4424733782603d355272b53fe91fcc4f67d60ed2edaa180f266be5aR181-R182) have an indentation of 24 spaces. However, we have no strict rules about style that so it's fine to keep the 4-space indentation here. Our only guideline is to try to keep formatting within a module consistent, but even that is not a strict rule

Reply:
> I'm not really sure what you mean by "... is not really a problem if you avoid those two indentations for one logical level". Could you elaborate?

Look at this code, for example:

https://github.com/IntersectMBO/lsm-tree/blob/2f5e9827fe2b8758fff32935628b2f3da7239a47/src/Database/LSMTree/Internal/IndexCompact.hs#L882-L887

The `where` is indented relative to the definition of `bitIndexFromTo`, and the local definitions are indented relative to the `where`, although they are just subordinate to `bitIndexFromTo` and thus logically one level below it. If you would use this indentation style with 4 spaces, you would indent the local definitions 8 spaces relative to the definition of `bitIndexFromTo`, which would be problematic.

> IMO 2-space indentation saves a lot of horizontal real estate in source code, and that's why I prefer it.

Yes, the use of horizontal space is an issue, particularly with Haskell, where code tends to be wider than with typical imperative languages. It’s a tradeoff between readability and economy. Actually, a while ago I came to the conclusion that 3-space indentation might be a good compromise but decided against using it, also in my personal projects, because it seemed too exotic (although it’s part of the standard style for Ada, the programming language).

I think I’ll stick to 4-space indentation for now.